File: apps/web/src/services/transformations.ts
import { apiGet, apiPatch, apiPost } from "@/services/api-client";
import type {
  DatasetVersion,
  DatasetVersionDetail,
  Lineage,
  TransformationApplyResult,
  TransformationDefinition,
  TransformationPreview,
  TransformationRequest,
  VersionComparison,
  WorkspaceStats,
} from "@/types/transformations";
import type { DatasetPreview } from "@/types/dataset";
import type { DatasetProfile } from "@/types/profile";

const TRANSFORM_TIMEOUT_MS = 120_000;

export function listTransformations(): Promise<TransformationDefinition[]> {
  return apiGet<TransformationDefinition[]>("/api/v1/transformations");
}

export function listVersions(datasetId: string): Promise<DatasetVersion[]> {
  return apiGet<DatasetVersion[]>(`/api/v1/datasets/${datasetId}/versions`);
}

export function getVersion(
  datasetId: string,
  versionId: string,
): Promise<DatasetVersionDetail> {
  return apiGet<DatasetVersionDetail>(
    `/api/v1/datasets/${datasetId}/versions/${versionId}`,
  );
}

export function getVersionPreview(
  datasetId: string,
  versionId: string,
): Promise<DatasetPreview> {
  return apiGet<DatasetPreview>(
    `/api/v1/datasets/${datasetId}/versions/${versionId}/preview`,
  );
}

export function getVersionProfile(
  datasetId: string,
  versionId: string,
): Promise<DatasetProfile> {
  return apiGet<DatasetProfile>(
    `/api/v1/datasets/${datasetId}/versions/${versionId}/profile`,
  );
}

export function previewTransformation(
  datasetId: string,
  versionId: string,
  body: TransformationRequest,
): Promise<TransformationPreview> {
  return apiPost<TransformationPreview>(
    `/api/v1/datasets/${datasetId}/versions/${versionId}/transformations/preview`,
    body,
    TRANSFORM_TIMEOUT_MS,
  );
}

export function applyTransformation(
  datasetId: string,
  versionId: string,
  body: TransformationRequest,
): Promise<TransformationApplyResult> {
  return apiPost<TransformationApplyResult>(
    `/api/v1/datasets/${datasetId}/versions/${versionId}/transformations`,
    body,
    TRANSFORM_TIMEOUT_MS,
  );
}

export function getLineage(datasetId: string, versionId: string): Promise<Lineage> {
  return apiGet<Lineage>(`/api/v1/datasets/${datasetId}/versions/${versionId}/lineage`);
}

export function getVersionComparison(
  datasetId: string,
  versionId: string,
): Promise<VersionComparison> {
  return apiGet<VersionComparison>(
    `/api/v1/datasets/${datasetId}/versions/${versionId}/comparison`,
  );
}

export function setCurrentVersion(
  datasetId: string,
  versionId: string,
): Promise<DatasetVersionDetail> {
  return apiPatch<DatasetVersionDetail>(`/api/v1/datasets/${datasetId}/current-version`, {
    version_id: versionId,
  });
}

export function getWorkspaceStats(): Promise<WorkspaceStats> {
  return apiGet<WorkspaceStats>("/api/v1/workspace/summary");
}

File: apps/api/src/facilio/api/v1/transformations.py
"""Transformation catalog, versions, preview, apply, and lineage."""

from __future__ import annotations

from flask import Blueprint, current_app, request

from facilio.core.responses import success_response
from facilio.db.session import Database
from facilio.schemas.transformations import (
    SetCurrentVersionRequest,
    TransformationRequest,
)
from facilio.services.profiles import ProfileService
from facilio.services.transformations import TransformationService


def _service() -> TransformationService:
    settings = current_app.config["FACILIO_SETTINGS"]
    database: Database = current_app.extensions["database"]
    return TransformationService(settings, database)


def _profiles() -> ProfileService:
    settings = current_app.config["FACILIO_SETTINGS"]
    database: Database = current_app.extensions["database"]
    return ProfileService(settings, database)


def register(blueprint: Blueprint) -> None:
    @blueprint.get("/transformations")
    def transformation_catalog():
        return success_response(_service().catalog())

    @blueprint.get("/workspace/summary")
    def workspace_summary():
        return success_response(_service().workspace_stats().model_dump(mode="json"))

    @blueprint.get("/datasets/<dataset_id>/versions")
    def list_versions(dataset_id: str):
        items = _service().list_versions(dataset_id)
        return success_response([item.model_dump(mode="json") for item in items])

    @blueprint.get("/datasets/<dataset_id>/versions/<version_id>")
    def get_version(dataset_id: str, version_id: str):
        payload = _service().get_version(dataset_id, version_id)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.get("/datasets/<dataset_id>/versions/<version_id>/preview")
    def preview_version(dataset_id: str, version_id: str):
        payload = _service().preview_version(dataset_id, version_id)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.get("/datasets/<dataset_id>/versions/<version_id>/profile")
    def get_version_profile(dataset_id: str, version_id: str):
        payload = _profiles().get_profile(dataset_id, version_id)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.post("/datasets/<dataset_id>/versions/<version_id>/profile")
    def run_version_profile(dataset_id: str, version_id: str):
        payload = _profiles().run_profile(dataset_id, version_id)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.get("/datasets/<dataset_id>/versions/<version_id>/lineage")
    def lineage(dataset_id: str, version_id: str):
        payload = _service().lineage(dataset_id, version_id)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.get("/datasets/<dataset_id>/versions/<version_id>/comparison")
    def comparison(dataset_id: str, version_id: str):
        payload = _service().compare_with_parent(dataset_id, version_id)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.post(
        "/datasets/<dataset_id>/versions/<version_id>/transformations/preview"
    )
    def preview_transformation(dataset_id: str, version_id: str):
        body = TransformationRequest.model_validate(request.get_json(silent=True) or {})
        payload = _service().preview_transformation(dataset_id, version_id, body)
        return success_response(payload.model_dump(mode="json"))

    @blueprint.post("/datasets/<dataset_id>/versions/<version_id>/transformations")
    def apply_transformation(dataset_id: str, version_id: str):
        body = TransformationRequest.model_validate(request.get_json(silent=True) or {})
        payload = _service().apply_transformation(dataset_id, version_id, body)
        return success_response(payload.model_dump(mode="json"), status_code=201)

    @blueprint.patch("/datasets/<dataset_id>/current-version")
    def set_current_version(dataset_id: str):
        body = SetCurrentVersionRequest.model_validate(
            request.get_json(silent=True) or {}
        )
        payload = _service().set_current_version(dataset_id, body)
        return success_response(payload.model_dump(mode="json"))
